### ISSUE TITLE
fish_preexec is no longer triggered on empty commands, no way to restore old behavior

### ISSUE THREAD

**Issue by gentoo-root** [fish-shell/fish-shell#8020]
I'm using fish 3.2.2 on linux.

I have a custom config under /etc/fish/conf.d to sync history across fish instances:

```
if status --is-interactive
	function sync_history --on-event fish_preexec
		history --save
		history --merge
	end
end
```

It used to work, but after #7085 empty commands no longer trigger the fish_preexec event, breaking my [workflow](https://imgs.xkcd.com/comics/workflow.png).

The old behavior was very convenient for me in various use cases:

1. I run command A in terminal 1, then commands B, C, D in terminal 2, while terminal 1 is idling. By pressing `<up>, <enter>` in terminal 1, I could rerun command A (the last command in that terminal).
2. I run command A in terminal 2, then commands B, C, D in terminal 2. By pressing `<enter>` in terminal 1, I could sync history and get commands B, C, D available in terminal 1.
3. I run a long-running command A in terminal 1. In the meanwhile, I run commands B, C, D in terminal 2, then command A finishes execution. Commands B, C, D didn't sync to terminal 1 automatically, so I could choose: either I could rerun command A by pressing `<up>, <enter>` in terminal 1, or I could press `<enter>` in terminal 1 and get the full history.

The summary of the old behavior is that any time I saw the prompt, I could safely assume that the last command I saw on the screen would be the one executed by `<up>, <enter>`, and I could easily sync history by pressing <enter>.

With this behavior change, I no longer get history updates, unless I run a non-empty command. If I replace fish_preexec with fish_prompt, I will get history updates, but lose the ability to safely and blindly press `<up>, <enter>` to rerun the last command in this terminal. According to the [docs](https://fishshell.com/docs/current/cmds/function.html?highlight=fish_preexec), there is no alternative event, which would get triggered *before* empty and non-empty commands.

I understand there could be a motivation to break backward compatibility of fish_preexec, but I would like to have at least another event I could migrate to.

Note it's not a dup of #7261, because the reporter found a workaround for fish_postexec, while I need a solution or workaround for fish_preexec.

**Comment by faho** [comment]
>With this behavior change, I no longer get history updates, unless I run a non-empty command. If I replace fish_preexec with fish_prompt, I will get history updates, but lose the ability to safely and blindly press <up>, <enter> to rerun the last command in this terminal. 

I don't see why that would be?

Before, the events run on empty commands:

- fish_preexec
- fish_postexec
- fish_prompt

Now, the events run on empty commands:

- fish_prompt

**Comment by gentoo-root** [comment]
> I don't see why that would be?

If I use fish_postexec or fish_prompt, then the following flow is possible:

1. I run a long-running command A in terminal 1.
2. I switch to terminal 2 and run commands B, C, D.
3. Command A finishes.
4. fish_postexec/fish_prompt hook executes and syncs the history in terminal 1.
5. I press `<up>, <enter>` in terminal 1.
6. Command D executes instead of command A. That's not what I'd expect.

The only option left is fish_preexec, which used to work best for me. However, after the behavior change the following flow no longer works:

1. I run a command A in terminal 1.
2. I switch to terminal 2 and run commands B, C, D.
3. I press `<enter>` in terminal 1 and get commands A, B, C, D in the history of terminal 1. This no longer works because empty commands skip fish_preexec.

> instead of repurposing the execute key simply bind something to sync history explicity - e.g. bind \cr 'history save; history merge'

That binds Ctrl+R to sync the history, right? Sorry, that's not what I want at all.

I got used to my old workflow and find it very convenient. This is a backward-incompatible change that led to a regression — why did you mark the ticket as a question?

**Comment by faho** [comment]
>This is a backward-incompatible change that led to a regression — why did you mark the ticket as a question?

Because it's not something we wish to re-implement. We don't want another "fish_preexec_but_with_empty_commandline" event or to go back to the previous state, so it's about how you can bend your config to do what you want. That's a question, not a bug or enhancement.

>That binds Ctrl+R to sync the history, right? Sorry, that's not what I want at all.

It is, and I would argue that's a much better experience.

If you really need enter, then rebind that - 

```fish
function sync-and-execute
    set -l cmd (commandline)
    if test -z "$cmd"
        history save
        history merge
    end
    commandline -f execute
end

bind \r sync-and-execute
```

or something similar.

**Comment by faho** [comment]
>Really?

Sorry, no, that was meant to be "it isn't". I was distracted and broke the comment in the edit.

>Sorry, but that's some utterly broken logic. It is a bug, it falls exactly under this definition:

It doesn't, because it's on purpose.

We don't want a new event, we don't want to trigger the event again.

> I can work on a patch

Sorry, I don't want to take a patch for this. Either we find a way for you to configure it how you want, or we don't find anything.

**Comment by faho** [comment]
Alright, question has been answered, closing!

**Comment by gentoo-root** [comment]
>Hence the test -z check.

That was something that I missed. Now I understand how it's supposed to work, thanks for this workaround. For non-empty commands I get fish_preexec, and for empty ones we fall into this branch.

I also have a postexec hook that stopped working on empty commands — I'll see if I can work around that with fish_prompt.

On the other hand, I still see other issues with that change (#7085). There are whitespace-only commands, which are treated as empty by $status, history and $CMD_DURATION (well, for history it may be because all commands starting from space are ignored), but fish_preexec is still emitted for them. So, the inconsistency still exists. To be honest, I don't want it to be fixed, because it would break the `test -z` check, which will need to be replaced by something else, but I realize that it may be discovered sooner or later anyway. It would be nice to have a builtin function that tells if fish is going to treat the command as empty or not, it would allow to share the same logic between fish internals and user scripts.

**Comment by faho** [comment]
Just noticed there's a simpler way to accomplish this - just send fish_preexec yourself!

```fish
function execute-or-preexec
    set -l cmd (commandline)
    if test -z "$cmd"
         emit fish_preexec
    end
    commandline -f execute
end

bind \r execute-or-preexec
```

**Comment by gentoo-root** [comment]
>just send fish_preexec yourself

Cool, thanks for the tip! I was actually just calling my sync_history function, so I had proper code sharing, but emitting the signal directly is closer to restore the old behavior.

What do you think about the last paragraph (about whitespace-only commands)?

**Comment by dpc** [comment]
The `bind \r` trick doesn't work.

**Comment by krobelus** [comment]
> The bind \r trick doesn't work.

Works for me.

If you use Vi mode, you might want to also bind it in other modes
such as insert:

	bind -M insert \r execute-or-preexec

If you are pressing ctrl-j or ctrl-m instead of `enter`, those are different
keys so need to be bound separately on capable terminals.

	bind ctrl-j execute-or-preexec
	bind -M insert ctrl-j execute-or-preexec
	bind ctrl-m execute-or-preexec
	...

**Comment by dpc** [comment]
Oh, OK. Thanks. `--preset` was breaking it.
